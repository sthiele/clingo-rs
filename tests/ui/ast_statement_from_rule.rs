use clingo::*;

fn main() {
    let sym = Symbol::create_id("test", true).unwrap();
    let mut term1 = ast::Term::from(sym);
    let mut term2 = ast::Term::from(sym);
    let mut lit1 = ast::Literal::from_term(ast::Sign::NoSign, &term1);
    let mut lit2 = ast::Literal::from_term(ast::Sign::NoSign, &term2);
    let hlit = ast::HeadLiteral::from(&lit1);
    let blit = ast::BodyLiteral::from_literal(ast::Sign::NoSign, &lit2);
    let body = vec![blit];
    let mut rule = ast::Rule::new(hlit, &body);
    let stmt = rule.ast_statement();
    
    term1 = ast::Term::from(sym);
    term2 = ast::Term::from(sym);
    lit1 = ast::Literal::from_term(ast::Sign::NoSign, &term1);
    lit2 = ast::Literal::from_term(ast::Sign::NoSign, &term2);
    rule = ast::Rule::new(hlit, &body);
    drop(body);
    let _end = (lit1, lit2, rule, stmt);
}