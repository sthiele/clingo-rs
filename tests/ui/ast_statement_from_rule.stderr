error[E0506]: cannot assign to `term1` because it is borrowed
  --> $DIR/ast_statement_from_rule.rs:15:5
   |
7  |     let mut lit1 = ast::Literal::from_term(ast::Sign::NoSign, &term1);
   |                                                               ------ borrow of `term1` occurs here
...
15 |     term1 = ast::Term::from(sym);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ assignment to borrowed `term1` occurs here
...
19 |     rule = ast::Rule::new(hlit, &body);
   |                           ---- borrow later used here

error[E0506]: cannot assign to `term2` because it is borrowed
  --> $DIR/ast_statement_from_rule.rs:16:5
   |
8  |     let mut lit2 = ast::Literal::from_term(ast::Sign::NoSign, &term2);
   |                                                               ------ borrow of `term2` occurs here
...
16 |     term2 = ast::Term::from(sym);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ assignment to borrowed `term2` occurs here
...
19 |     rule = ast::Rule::new(hlit, &body);
   |                                 ----- borrow later used here

error[E0506]: cannot assign to `lit1` because it is borrowed
  --> $DIR/ast_statement_from_rule.rs:17:5
   |
9  |     let hlit = ast::HeadLiteral::from(&lit1);
   |                                       ----- borrow of `lit1` occurs here
...
17 |     lit1 = ast::Literal::from_term(ast::Sign::NoSign, &term1);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ assignment to borrowed `lit1` occurs here
18 |     lit2 = ast::Literal::from_term(ast::Sign::NoSign, &term2);
19 |     rule = ast::Rule::new(hlit, &body);
   |                           ---- borrow later used here

error[E0506]: cannot assign to `lit2` because it is borrowed
  --> $DIR/ast_statement_from_rule.rs:18:5
   |
10 |     let blit = ast::BodyLiteral::from_literal(ast::Sign::NoSign, &lit2);
   |                                                                  ----- borrow of `lit2` occurs here
...
18 |     lit2 = ast::Literal::from_term(ast::Sign::NoSign, &term2);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ assignment to borrowed `lit2` occurs here
19 |     rule = ast::Rule::new(hlit, &body);
   |                                 ----- borrow later used here

error[E0506]: cannot assign to `rule` because it is borrowed
  --> $DIR/ast_statement_from_rule.rs:19:5
   |
13 |     let stmt = rule.ast_statement();
   |                ---- borrow of `rule` occurs here
...
19 |     rule = ast::Rule::new(hlit, &body);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ assignment to borrowed `rule` occurs here
20 |     drop(body);
21 |     let _end = (lit1, lit2, rule, stmt);
   |                                   ---- borrow later used here

error[E0505]: cannot move out of `body` because it is borrowed
  --> $DIR/ast_statement_from_rule.rs:20:10
   |
12 |     let mut rule = ast::Rule::new(hlit, &body);
   |                                         ----- borrow of `body` occurs here
...
20 |     drop(body);
   |          ^^^^ move out of `body` occurs here
21 |     let _end = (lit1, lit2, rule, stmt);
   |                             ---- borrow later used here
